middleware.ts

```
@@ -0,0 +1,79 @@
+    
+    // Check session
+    const { data: { session } } = await supabase.auth.getSession();
+    
+    // Handle API routes
+    if (request.nextUrl.pathname.startsWith('/api/')) {
+      if (!session) {
+        return NextResponse.json({ error: 'Unauthorized' }, { status: 401 });
+      }
+      return res;
+    }
+
+    // Handle page routes
+    if (!session) {
+      // Allow access to login page
+      if (request.nextUrl.pathname === '/login') {
+        return res;
+      }
+      // Redirect to login for all other pages
+      return NextResponse.redirect(new URL('/login', request.url));
+    }
+
+    // Get user role from users table
+    const { data: userData, error: userError } = await supabase
+      .from('users')
+      .select('role')
+      .eq('id', session.user.id)
+      .single();
+
+    if (userError || !userData) {
+      return NextResponse.redirect(new URL('/login', request.url));
+    }
+
+    const role = userData.role;
+
+    // Handle role-based routing
+    if (request.nextUrl.pathname === '/') {
+      // Redirect root to role-specific dashboard
+      return NextResponse.redirect(new URL(`/dashboard/${role}`, request.url));
+    }
+
+    if (request.nextUrl.pathname === '/login') {
+      // Redirect authenticated users to their dashboard
+      return NextResponse.redirect(new URL(`/dashboard/${role}`, request.url));
+    }
+
+    if (request.nextUrl.pathname.startsWith('/dashboard/admin')) {
+      if (role !== 'admin') {
+        return NextResponse.redirect(new URL('/dashboard/collaborator', request.url));
+      }
+    } else if (request.nextUrl.pathname.startsWith('/dashboard/collaborator')) {
+      if (role !== 'collaborator') {
+        return NextResponse.redirect(new URL('/dashboard/admin', request.url));
+      }
+    }
+
+    return res;
+  } catch (error) {
+    console.error('Middleware error:', error);
+    return NextResponse.redirect(new URL('/login', request.url));
```

This can cause an infinite redirect if a middleware error occurs. I was running the project without the `.env` Supabase variables, and that caused the browser to quit the process due to too many redirects to `'/login'`. 

As we cannot predict the error that might occur, which may not necessarily be from a user accessing the wrong page, we can instead return an error response with a status code 500 and an error message.